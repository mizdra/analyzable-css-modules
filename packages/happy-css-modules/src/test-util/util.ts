import { constants, mkdirSync, realpathSync, rmSync, writeFileSync } from 'fs';
import { access } from 'fs/promises';
import { tmpdir } from 'os';
import { dirname, join, resolve } from 'path';
import postcss, { type Root, type Rule, type AtRule } from 'postcss';
import { type ClassName } from 'postcss-selector-parser';
import { type Token, collectNodes, type Location } from '../locator/index.js';
import { sleepSync } from '../util.js';

export const FIXTURE_DIR_PATH = resolve(
  realpathSync(tmpdir()),
  'happy-css-modules/fixtures',
  process.env['VITEST_WORKER_ID']!,
);

export function createRoot(code: string, from?: string): Root {
  return postcss.parse(code, { from: from || '/test/test.css' });
}

export function createAtImports(root: Root): AtRule[] {
  return collectNodes(root).atImports;
}

export function createAtValues(root: Root): AtRule[] {
  return collectNodes(root).atValues;
}

export function createClassSelectors(root: Root): { rule: Rule; classSelector: ClassName }[] {
  return collectNodes(root).classSelectors;
}

export function fakeToken(args: {
  name: Token['name'];
  originalLocation: { filePath?: Location['filePath']; start?: Location['start'] };
}): Token {
  if (args.originalLocation.filePath === undefined || args.originalLocation.start === undefined) {
    return {
      name: args.name,
      originalLocation: { filePath: undefined, start: undefined, end: undefined },
    };
  } else {
    return {
      name: args.name,
      originalLocation: {
        filePath: args.originalLocation.filePath ?? getFixturePath('/test/1.css'),
        start: args.originalLocation.start,
        end: {
          line: args.originalLocation.start.line,
          column: args.originalLocation.start.column + args.name.length - 1,
        },
      },
    };
  }
}

export async function waitForAsyncTask(ms?: number): Promise<void> {
  await new Promise((resolve) => {
    setTimeout(resolve, ms ?? 0);
  });
}

export async function exists(path: string): Promise<boolean> {
  try {
    await access(path, constants.F_OK);
    return true;
  } catch (e) {
    return false;
  }
}

type File = string;

type DirectoryItem = File | DirectoryItems;

type DirectoryItems = {
  [name: string]: DirectoryItem;
};

function isFile(item: DirectoryItem): item is File {
  return typeof item === 'string';
}

export function createFixtures(items: DirectoryItems): void {
  function createFixturesImpl(items: DirectoryItems, baseDir: string): void {
    for (const [name, item] of Object.entries(items)) {
      const path = join(baseDir, name);
      if (isFile(item)) {
        mkdirSync(dirname(path), { recursive: true });
        if (typeof item === 'string') {
          writeFileSync(path, item);
        }
      } else {
        mkdirSync(path, { recursive: true });
        createFixturesImpl(item, path);
      }
    }
  }
  removeFixtures();
  sleepSync(2); // Wait 2 ms for mtime to change from the previous fixture.
  createFixturesImpl(items, FIXTURE_DIR_PATH);
}

export function removeFixtures(): void {
  rmSync(FIXTURE_DIR_PATH, { recursive: true, force: true });
}

export function getFixturePath(path: string): string {
  return join(FIXTURE_DIR_PATH, path);
}
