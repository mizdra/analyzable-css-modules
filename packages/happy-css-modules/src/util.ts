import { constants, readFileSync } from 'fs';
import { access } from 'fs/promises';
import { join, dirname, resolve } from 'path';
import { fileURLToPath } from 'url';
import { resolve as importMetaResolve } from 'import-meta-resolve';
import minimatch from 'minimatch';

/**
 * The SystemError type of Node.js.
 * @see https://nodejs.org/api/errors.html#class-systemerror
 */
export interface SystemError {
  code: string;
}

export function isSystemError(value: unknown): value is SystemError {
  return (
    isObject(value) &&
    hasProp(value, 'constructor') &&
    isObject(value.constructor) &&
    hasProp(value.constructor, 'name') &&
    value.constructor.name === 'Error' &&
    hasProp(value, 'code') &&
    typeof value.code === 'string'
  );
}

export function isObject(value: unknown): value is object {
  return (typeof value === 'object' || typeof value === 'function') && value !== null;
}

export function hasProp<T extends string>(obj: object, prop: T): obj is { [key in T]: unknown } {
  return prop in obj;
}

export function unique<T>(array: T[]): T[] {
  return [...new Set(array)];
}

export function uniqueBy<T, U>(arr: T[], fn: (el: T) => U): T[] {
  const result: T[] = [];
  const keys = new Set<U>();
  for (const el of arr) {
    const key = fn(el);
    if (!keys.has(key)) {
      keys.add(key);
      result.push(el);
    }
  }
  return result;
}

export function sleepSync(ms: number) {
  const start = Date.now();
  // eslint-disable-next-line no-empty
  while (Date.now() - start < ms) {}
}

export async function exists(path: string): Promise<boolean> {
  try {
    await access(path, constants.F_OK);
    return true;
  } catch (_e) {
    return false;
  }
}

export function isMatchByGlob(filePath: string, pattern: string, options: { cwd: string }): boolean {
  return minimatch(filePath, join(options.cwd, pattern));
}

export function getPackageJson() {
  return JSON.parse(readFileSync(resolve(dirname(fileURLToPath(import.meta.url)), '../package.json'), 'utf-8'));
}

export async function getInstalledPeerDependencies(): Promise<string[]> {
  const pkgJson = getPackageJson();
  const result = [];
  for (const deps of Object.keys(pkgJson.peerDependencies)) {
    const isInstalled = await importMetaResolve(deps, import.meta.url)
      .then(() => true)
      .catch(() => false);
    if (isInstalled) result.push(deps);
  }
  return result;
}
