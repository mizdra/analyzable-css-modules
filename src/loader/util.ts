import { stat, readFile } from 'fs/promises';
import { createHash } from 'node:crypto';
import fetch from 'node-fetch';

// TODO: `--cache-strategy` option will allow users to switch between `content` and `metadata` modes.
// TODO: Implement a cache algorithm compliant with RFC 9111 and RFC 5861 for performance.

const REVISION_HEADER_NAMES = ['etag', 'last-modified'];

export async function fetchRevision(url: string): Promise<string> {
  const { protocol, href, pathname } = new URL(url);
  if (protocol === 'http:' || protocol === 'https:') {
    const res = await fetch(href, { method: 'GET' });
    if (!res.ok) throw new Error(`Failed to fetch ${href}: ${res.status} ${res.statusText}`);
    for (const headerName of REVISION_HEADER_NAMES) {
      const headerValue = res.headers.get(headerName);
      if (headerValue) return `${headerName}:${headerValue}`;
    }
    const content = await res.text();
    const contentHash = createHash('sha256').update(content).digest('hex');
    return `content-hash:${contentHash}`;
  } else if (protocol === 'file:') {
    const mtime = (await stat(pathname)).mtime.getTime();
    return `mtime:${mtime}`;
  } else {
    throw new Error(`Unsupported protocol: ${href}`);
  }
}

export async function fetchContent(url: string): Promise<string> {
  const { protocol, href, pathname } = new URL(url);
  if (protocol === 'http:' || protocol === 'https:') {
    return fetch(href).then(async (response) => response.text());
  } else if (protocol === 'file:') {
    return readFile(pathname, 'utf8');
  } else {
    throw new Error(`Unsupported protocol: ${url}`);
  }
}

export function isURL(url: string): boolean {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}

// export function isSupportedProtocol(url: string): boolean {
//   const { protocol } = new URL(url);
//   return protocol === 'file:' || protocol === 'http:' || protocol === 'https:';
// }
