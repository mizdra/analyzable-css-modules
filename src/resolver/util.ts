/**
 * Check whether a specifier is an absolute URL
 */
export function isAbsoluteURL(specifier: string): boolean {
  let url: URL;
  try {
    // Check if the URL is valid
    url = new URL(specifier);
  } catch (e) {
    // If the URL is invalid, return false.
    return false;
  }
  // Support file://, http://, https:// URLs only.
  return url.protocol === 'file:' || url.protocol === 'http:' || url.protocol === 'https:';
}
