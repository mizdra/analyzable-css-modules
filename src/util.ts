import { constants } from 'fs';
import { access, stat, readFile } from 'fs/promises';
import { createHash } from 'node:crypto';
import { join } from 'path';
import minimatch from 'minimatch';
import fetch from 'node-fetch';

const REVISION_HEADER_NAMES = ['etag', 'last-modified'];

/**
 * The SystemError type of Node.js.
 * @see https://nodejs.org/api/errors.html#class-systemerror
 */
export interface SystemError {
  code: string;
}

export function isSystemError(value: unknown): value is SystemError {
  return (
    isObject(value) &&
    hasProp(value, 'constructor') &&
    isObject(value.constructor) &&
    hasProp(value.constructor, 'name') &&
    value.constructor.name === 'Error' &&
    hasProp(value, 'code') &&
    typeof value.code === 'string'
  );
}

export function isObject(value: unknown): value is object {
  return (typeof value === 'object' || typeof value === 'function') && value !== null;
}

export function hasProp<T extends string>(obj: object, prop: T): obj is { [key in T]: unknown } {
  return prop in obj;
}

export function unique<T>(array: T[]): T[] {
  return [...new Set(array)];
}

export function uniqueBy<T, U>(arr: T[], fn: (el: T) => U): T[] {
  const result: T[] = [];
  const keys = new Set<U>();
  for (const el of arr) {
    const key = fn(el);
    if (!keys.has(key)) {
      keys.add(key);
      result.push(el);
    }
  }
  return result;
}

export function sleepSync(ms: number) {
  const start = Date.now();
  // eslint-disable-next-line no-empty
  while (Date.now() - start < ms) {}
}

export async function exists(path: string): Promise<boolean> {
  try {
    await access(path, constants.F_OK);
    return true;
  } catch (e) {
    return false;
  }
}

export function isMatchByGlob(filePath: string, pattern: string, options: { cwd: string }): boolean {
  return minimatch(filePath, join(options.cwd, pattern));
}

export async function fetchRevision(url: string): Promise<string> {
  // TODO: `--cache-strategy` option will allow users to switch between `content` and `metadata` modes.
  // TODO: Implement a cache algorithm compliant with RFC 9111 and RFC 5861 for performance.

  const { protocol, href, pathname } = new URL(url);
  if (protocol === 'http:' || protocol === 'https:') {
    const res = await fetch(href, { method: 'GET' });
    if (!res.ok) throw new Error(`Failed to fetch ${href}: ${res.status} ${res.statusText}`);
    for (const headerName of REVISION_HEADER_NAMES) {
      const headerValue = res.headers.get(headerName);
      if (headerValue) return `${headerName}:${headerValue}`;
    }
    const content = await res.text();
    const contentHash = createHash('sha256').update(content).digest('hex');
    return `content-hash:${contentHash}`;
  } else if (protocol === 'file:') {
    const mtime = (await stat(pathname)).mtime.getTime();
    return `mtime:${mtime}`;
  } else {
    throw new Error(`Unsupported protocol: ${href}`);
  }
}

export async function fetchContent(url: string): Promise<string> {
  // TODO: Implement a cache algorithm compliant with RFC 9111 and RFC 5861 for performance.

  const { protocol, href, pathname } = new URL(url);
  if (protocol === 'http:' || protocol === 'https:') {
    return fetch(href).then(async (response) => response.text());
  } else if (protocol === 'file:') {
    return readFile(pathname, 'utf8');
  } else {
    throw new Error(`Unsupported protocol: ${url}`);
  }
}

export function isURL(url: string): boolean {
  try {
    new URL(url);
    return true;
  } catch {
    return false;
  }
}
