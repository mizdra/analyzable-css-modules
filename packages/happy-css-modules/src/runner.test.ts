import { readFile, rm, symlink, writeFile } from 'fs/promises';
import { randomUUID } from 'node:crypto';
import { createRequire } from 'node:module';
import { dirname, join, resolve } from 'path';
import { fileURLToPath } from 'url';
import * as fileCacheNpm from '@file-cache/npm';
import { jest } from '@jest/globals';
import dedent from 'dedent';
import type { RunnerOptions, Watcher } from './runner.js';
import { createFixtures, exists, getFixturePath, waitForAsyncTask } from './test-util/util.js';

const packageRootDir = resolve(dirname(fileURLToPath(import.meta.url)), '..');

const require = createRequire(import.meta.url);

jest.unstable_mockModule('@file-cache/npm', () => ({
  ...fileCacheNpm, // Inherit native functions
  createNpmPackageKey: () => 'mocked-key',
}));

jest.unstable_mockModule('pkg-dir', () => ({
  packageDirectory: () => packageRootDir,
}));

const { run } = await import('./runner.js');

// eslint-disable-next-line @typescript-eslint/no-empty-function
const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
// eslint-disable-next-line @typescript-eslint/no-empty-function
const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

const defaultOptions: RunnerOptions = {
  pattern: 'test/**/*.{css,scss}',
  declarationMap: true,
  logLevel: 'silent',
  cwd: getFixturePath('/'),
  cache: false,
};

const dir = join(dirname(fileURLToPath(import.meta.url)));

beforeEach(async () => {
  consoleLogSpy.mockClear();
  consoleErrorSpy.mockClear();
  await rm(resolve(dir, '../node_modules/.cache/happy-css-modules'), { recursive: true, force: true }); // clear cache
});

// Exit the watcher even if the test fails
let watcher: Watcher | undefined;
afterEach(async () => {
  if (watcher) {
    await watcher.close();
  }
});

test('generates .d.ts and .d.ts.map', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
    '/test/2.css': '.b {}',
  });
  await run({ ...defaultOptions });
  expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/1.css.d.ts.map'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/2.css.d.ts'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/2.css.d.ts.map'), 'utf8')).toMatchSnapshot();
});

test('uses cache in non-watch mode', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
  });
  await run({ ...defaultOptions, declarationMap: true, logLevel: 'debug', cache: true });
  expect(consoleLogSpy).toBeCalledTimes(2);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(1, expect.anything(), expect.stringContaining('Generate .d.ts for'));
  expect(consoleLogSpy).toHaveBeenNthCalledWith(2, expect.anything(), expect.stringContaining('generated'));
  consoleLogSpy.mockClear();

  // Skip generation
  await run({ ...defaultOptions, declarationMap: true, logLevel: 'debug', cache: true });
  expect(consoleLogSpy).toBeCalledTimes(2);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(1, expect.anything(), expect.stringContaining('Generate .d.ts for'));
  expect(consoleLogSpy).toHaveBeenNthCalledWith(2, expect.anything(), expect.stringContaining('skipped'));
  consoleLogSpy.mockClear();

  // Generates if generated files are missing
  await rm(getFixturePath('/test/1.css.d.ts'));
  await run({ ...defaultOptions, declarationMap: true, logLevel: 'debug', cache: true });
  expect(consoleLogSpy).toBeCalledTimes(2);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(1, expect.anything(), expect.stringContaining('Generate .d.ts for'));
  expect(consoleLogSpy).toHaveBeenNthCalledWith(2, expect.anything(), expect.stringContaining('generated'));
  consoleLogSpy.mockClear();

  // Generates if options are changed
  await run({ ...defaultOptions, declarationMap: false, logLevel: 'debug', cache: true });
  expect(consoleLogSpy).toBeCalledTimes(2);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(1, expect.anything(), expect.stringContaining('Generate .d.ts for'));
  expect(consoleLogSpy).toHaveBeenNthCalledWith(2, expect.anything(), expect.stringContaining('generated'));
  consoleLogSpy.mockClear();
});

test('uses cache in watch mode', async () => {
  createFixtures({
    '/test/1.css': '.a-1 {}',
    '/test/2.css': '.b-1 {}',
  });

  // At first, process all files
  watcher = await run({ ...defaultOptions, declarationMap: true, logLevel: 'debug', cache: true, watch: true });
  await waitForAsyncTask(1000); // Wait until the watcher is ready
  expect(consoleLogSpy).toBeCalledTimes(3);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(
    1,
    expect.anything(),
    expect.stringContaining('Watch test/**/*.{css,scss}...'),
  );
  expect(consoleLogSpy).toHaveBeenNthCalledWith(
    2,
    expect.anything(),
    expect.stringContaining('test/1.css (generated)'),
  );
  expect(consoleLogSpy).toHaveBeenNthCalledWith(
    3,
    expect.anything(),
    expect.stringContaining('test/2.css (generated)'),
  );
  consoleLogSpy.mockClear();

  // Updating 1.css, it will only be processed
  await writeFile(getFixturePath('/test/1.css'), '.a-2 {}');
  await waitForAsyncTask(500); // Wait until the file is written
  expect(consoleLogSpy).toBeCalledTimes(1);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(
    1,
    expect.anything(),
    expect.stringContaining('test/1.css (generated)'),
  );

  // Close the watcher
  await watcher.close();
  consoleLogSpy.mockClear();

  // Update 1.css
  await writeFile(getFixturePath('/test/1.css'), '.a-1 {}');
  await waitForAsyncTask(500); // Wait until the file is written

  // The updated 1.css will be processed, and the non-updated 2.css will be skipped.
  watcher = await run({ ...defaultOptions, declarationMap: true, logLevel: 'debug', cache: true, watch: true });
  await waitForAsyncTask(1000); // Wait until the watcher is ready
  expect(consoleLogSpy).toBeCalledTimes(3);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(
    1,
    expect.anything(),
    expect.stringContaining('Watch test/**/*.{css,scss}...'),
  );
  expect(consoleLogSpy).toHaveBeenNthCalledWith(
    2,
    expect.anything(),
    expect.stringContaining('test/1.css (generated)'),
  );
  expect(consoleLogSpy).toHaveBeenNthCalledWith(3, expect.anything(), expect.stringContaining('test/2.css (skipped)'));
});

test('outputs logs', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
  });
  await run({ ...defaultOptions, logLevel: 'debug', cache: true });
  expect(consoleLogSpy).toBeCalledTimes(2);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(1, expect.anything(), expect.stringContaining('Generate .d.ts for'));
  expect(consoleLogSpy).toHaveBeenNthCalledWith(
    2,
    expect.anything(),
    expect.stringContaining('test/1.css (generated)'),
  );
  consoleLogSpy.mockClear();

  await run({ ...defaultOptions, logLevel: 'debug', cache: true });
  expect(consoleLogSpy).toBeCalledTimes(2);
  expect(consoleLogSpy).toHaveBeenNthCalledWith(1, expect.anything(), expect.stringContaining('Generate .d.ts for'));
  expect(consoleLogSpy).toHaveBeenNthCalledWith(2, expect.anything(), expect.stringContaining('test/1.css (skipped)'));
});

test.todo('changes dts format with localsConvention options');
test('does not emit declaration map if declarationMap is false', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
  });
  await run({ ...defaultOptions, declarationMap: false });
  await expect(readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).resolves.not.toThrow();
  await expect(readFile(getFixturePath('/test/1.css.d.ts.map'), 'utf8')).rejects.toThrow(/ENOENT/u);
});
test('supports transformer', async () => {
  createFixtures({
    '/test/1.scss': `.a { dummy: ''; }`,
  });
  await run({ ...defaultOptions });
  expect(await readFile(getFixturePath('/test/1.scss.d.ts'), 'utf8')).toMatchSnapshot();
  expect(await readFile(getFixturePath('/test/1.scss.d.ts.map'), 'utf8')).toMatchSnapshot();
});
test('watches for changes in files', async () => {
  createFixtures({
    '/test': {}, // empty directory
  });
  watcher = await run({ ...defaultOptions, watch: true });

  await writeFile(getFixturePath('/test/1.css'), '.a-1 {}');
  await waitForAsyncTask(500); // Wait until the file is written
  expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatch(/a-1/u);

  await writeFile(getFixturePath('/test/1.css'), '.a-2 {}');
  await waitForAsyncTask(500); // Wait until the file is written
  expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatch(/a-2/u);

  await writeFile(getFixturePath('/test/2.css'), '.b {}');
  await writeFile(getFixturePath('/test/3.css'), '.c {}');
  await waitForAsyncTask(500); // Wait until the file is written
  expect(await readFile(getFixturePath('/test/2.css.d.ts'), 'utf8')).toMatch(/b/u);
  expect(await readFile(getFixturePath('/test/3.css.d.ts'), 'utf8')).toMatch(/c/u);
});
test('returns an error if the file fails to process in non-watch mode', async () => {
  createFixtures({
    '/test/1.css': '.a {}',
    '/test/2.css': 'INVALID SYNTAX',
    '/test/3.css': 'INVALID SYNTAX',
  });
  const maybeError = await run({ ...defaultOptions, watch: false }).catch((e) => e);

  // The errors are aggregated into AggregateError.
  expect(maybeError).toBeInstanceOf(AggregateError);
  const error = maybeError as AggregateError;
  expect(error.message).toMatchInlineSnapshot(`"Failed to process files"`);
  expect(error.errors).toHaveLength(2);
  expect(error.errors[0]).toMatchInlineSnapshot(`<fixtures>/test/2.css:1:1: Unknown word`);
  expect(error.errors[1]).toMatchInlineSnapshot(`<fixtures>/test/3.css:1:1: Unknown word`);

  // The valid files are emitted.
  expect(await exists(getFixturePath('/test/1.css.d.ts'))).toBe(true);
  expect(await exists(getFixturePath('/test/1.css.d.ts.map'))).toBe(true);
});
describe('handles external files', () => {
  beforeEach(() => {
    createFixtures({
      '/test/1.css': dedent`
      @import './2.css';
      @import 'external-library';
      .a {}
      `,
      '/test/2.css': `.b {}`,
      '/node_modules/external-library/index.css': `.c {}`,
    });
  });
  test('do not emit .dts for external files', async () => {
    await run({ ...defaultOptions });
    expect(await exists(getFixturePath('/test/1.css.d.ts'))).toBe(true);
    expect(await exists(getFixturePath('/test/2.css.d.ts'))).toBe(true);
    expect(await exists(getFixturePath('/node_modules/external-library/index.css.d.ts'))).toBe(false);
  });
  test('treats imported tokens from external files the same as local tokens', async () => {
    await run({ ...defaultOptions });
    expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatchInlineSnapshot(`
          "declare const styles:
            & Readonly<Pick<(typeof import("./2.css"))["default"], "b">>
            & Readonly<{ "c": string }>
            & Readonly<{ "a": string }>
          ;
          export default styles;
          //# sourceMappingURL=./1.css.d.ts.map
          "
      `);
  });
});

test('sassLoadPaths', async () => {
  const sassLoadPaths = ['test/relative'];
  createFixtures({
    '/test/1.scss': dedent`
    @import '2.scss';
    `,
    '/test/relative/2.scss': `.a { dummy: ''; }`,
  });
  await run({ ...defaultOptions, sassLoadPaths }); // not throw
});

test('lessIncludePaths', async () => {
  const lessIncludePaths = ['test/relative'];
  createFixtures({
    '/test/1.less': dedent`
    @import '2.less';
    `,
    '/test/relative/2.less': `.a { dummy: ''; }`,
  });
  await run({ ...defaultOptions, lessIncludePaths }); // not throw
});

test('webpackResolveAlias', async () => {
  const webpackResolveAlias = { '@relative': 'test/relative' };
  createFixtures({
    '/test/1.less': dedent`
    @import '@relative/2.less';
    `,
    '/test/relative/2.less': `.a { dummy: ''; }`,
  });
  await run({ ...defaultOptions, webpackResolveAlias }); // not throw
});

test('postcssConfig', async () => {
  const uuid = randomUUID();
  const postcssConfig = `${uuid}/postcss.config.js`;
  createFixtures({
    [`/${uuid}/postcss.config.js`]: dedent`
    module.exports = {
      plugins: [
        require('${require.resolve('postcss-simple-vars')}'),
      ],
    };
    `,
    '/test/1.css': dedent`
    $prefix: foo;
    .$(prefix)_bar {}
    `,
  });
  await run({ ...defaultOptions, postcssConfig }); // not throw
});

test('support symlink', async () => {
  createFixtures({
    '/external/1.css': '.a {}',
    '/external/2.css': '.b {}',
    '/test': {}, // empty directory
  });
  await symlink(getFixturePath('/external/1.css'), getFixturePath('/test/1.css'));
  await symlink(getFixturePath('/external/2.css'), getFixturePath('/test/2.txt'));

  await run({ ...defaultOptions, watch: false });

  // Symlinks that do not match the pattern are not processed.
  expect(await exists(getFixturePath('/test/1.css.d.ts'))).toBe(true);
  expect(await exists(getFixturePath('/test/2.css.d.ts'))).toBe(false);
  expect(await exists(getFixturePath('/test/2.txt.d.ts'))).toBe(false);

  // The path referred to by sourceMappingURL or sources field is the path before symlink resolution.
  expect(await readFile(getFixturePath('/test/1.css.d.ts'), 'utf8')).toMatchInlineSnapshot(`
    "declare const styles:
      & Readonly<{ "a": string }>
    ;
    export default styles;
    //# sourceMappingURL=./1.css.d.ts.map
    "
  `);
  expect(await readFile(getFixturePath('/test/1.css.d.ts.map'), 'utf8')).toMatchInlineSnapshot(
    `"{"version":3,"sources":["./1.css"],"names":["a"],"mappings":"AAAA;AAAA,E,aAAAA,G,WAAA;AAAA;AAAA","file":"1.css.d.ts","sourceRoot":""}"`,
  );
});
