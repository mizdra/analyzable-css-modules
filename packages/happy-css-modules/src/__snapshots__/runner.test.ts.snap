// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates .d.ts and .d.ts.map 1`] = `
"declare const styles:
  & Readonly<{ "a": string }>
;
export default styles;
//# sourceMappingURL=./1.css.d.ts.map
"
`;

exports[`generates .d.ts and .d.ts.map 2`] = `"{"version":3,"sources":["./1.css"],"names":["a"],"mappings":"AAAA;AAAA,E,aAAAA,G,WAAA;AAAA;AAAA","file":"1.css.d.ts","sourceRoot":""}"`;

exports[`generates .d.ts and .d.ts.map 3`] = `
"declare const styles:
  & Readonly<{ "b": string }>
;
export default styles;
//# sourceMappingURL=./2.css.d.ts.map
"
`;

exports[`generates .d.ts and .d.ts.map 4`] = `"{"version":3,"sources":["./2.css"],"names":["b"],"mappings":"AAAA;AAAA,E,aAAAA,G,WAAA;AAAA;AAAA","file":"2.css.d.ts","sourceRoot":""}"`;

exports[`supports transformer 1`] = `
"declare const styles:
  & Readonly<{ "a": string }>
;
export default styles;
//# sourceMappingURL=./1.scss.d.ts.map
"
`;

exports[`supports transformer 2`] = `"{"version":3,"sources":["./1.scss"],"names":["a"],"mappings":"AAAA;AAAA,E,aAAAA,G,WAAA;AAAA;AAAA","file":"1.scss.d.ts","sourceRoot":""}"`;
